### DISCUSSION TITLE
Hook into container build process

### DISCUSSION THREAD

**Discussion by mu88** [dotnet/sdk-container-builds#521] [Q&A] [answered]
Hi 👋🏻

First, I want to highlight how great your project is working 💪🏻 I migrated several of my apps from building the Docker image via `Dockerfile` to using the SDK Container Building Tools and it works super smooth.

When doing so, an idea/question came to my mind. With the current solution, I see no way to migrate this [`Dockerfile`](https://github.com/mu88/ScreenshotCreator/blob/main/src/ScreenshotCreator.Api/Dockerfile). While I might somehow could move the [custom PowerShell/Playwright installation logic](https://github.com/mu88/ScreenshotCreator/blob/main/src/ScreenshotCreator.Api/Dockerfile#L4-L20) into a dedicated image and use that as base image, [this line](https://github.com/mu88/ScreenshotCreator/blob/main/src/ScreenshotCreator.Api/Dockerfile#L31) will make me fall over the _favor SDK Container Building Tools over `Dockerfile`_ cliff 😉

To support such use cases, it would be cool to somehow hook into the container building process, allowing to add calls like `RUN ...`.

**Comment by baronfel** [comment]
Thanks for the kind words - it's always great to hear that folks are having success with the tooling :)

  You've hit the kind of situation of think about a lot - this idea of the "Dockerfile complexity cliff" is something we've really tried to address. For `RUN` commands we can't really address those for the fundamental reason that there is no Linux VM that we could run the commands in - though we do have an issue for [using a local Dockerfile as the base image for the running application](https://github.com/dotnet/sdk-container-builds/issues/125). 

I'm very curious about the `chmod` line you highlighted though - why did you need to add that to the container?

**Reply by mu88** [reply]
My app adds both content to a SQLite database and saves images to the filesystem. For this, I'm mounting a host directory into the container, containing the SQLite database and the directory where the images shall be saved.

To make this work, I had to `chown` the database to the `app` user's UID (which still feels a bit odd to me that a container user has to own the database on the host to read and write to it, but maybe I'm just not deep enough in the topic) and `chmod` the directory so that images can be saved.

**Reply by baronfel** [reply]
That's kind of counter to the purpose of using a non-root user - the goal is that nothing in the application's binaries' directory can be changed. Instead consider writing the database and saving the images to a directory that the app user has permissions to - a safe default would be `/home/app`. There's a workshop scenario demonstrating an example of how to do this over at [richlander/container-workshop](https://github.com/richlander/container-workshop/blob/main/super-sql-app/control-container-runtime.md) if you have time to take a look.

**Reply by mu88** [reply]
Sorry, I was writing stupid things and mixed up two different projects of mine 🤦🏻‍♂️ this project that I'm referring to creates screenshots of a website using Playwright. The Playwright driver however has to be started from the current application directory ([see here](https://github.com/microsoft/playwright-dotnet/blob/main/src/Playwright/Helpers/Driver.cs#L49)). So you're right that I can modify my app to not write the screenshots into the `/app` dir, but when omitting `RUN chmod -R 777 /app` I can no longer start my app due to the following error:
```
fail: Microsoft.Extensions.Hosting.Internal.Host[11]
      Hosting failed to start
      System.ComponentModel.Win32Exception (13): An error occurred trying to start process '/app/.playwright/node/linux-x64/playwright.sh' with working directory '/app'. Permission denied
         at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
         at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
         at Microsoft.Playwright.Transport.StdIOTransport.StartProcessWithUTF8IOEncoding(Process process) in /_/src/Playwright/Transport/StdIOTransport.cs:line 167
         at Microsoft.Playwright.Transport.StdIOTransport..ctor() in /_/src/Playwright/Transport/StdIOTransport.cs:line 48
         at Microsoft.Playwright.Playwright.CreateAsync() in /_/src/Playwright/Playwright.cs:line 44
         at ScreenshotCreator.Logic.PlaywrightHelper.InitializePlaywrightAsync() in /src/src/ScreenshotCreator.Logic/PlaywrightHelper.cs:line 43
```

And the remainder (i. e. PowerShell/Playwright installation stuff) you'd move into a dedicated image and build this via a `Dockerfile`?

Also thx for pointing me to the workshop scenario, that was helpful. I'd consider it even more helpful if you could add a section _Mount a SQLite database from the host into the container_ [here](https://github.com/richlander/container-workshop/blob/main/super-sql-app/control-container-runtime.md#change-your-application-to-react-to-non-root-permissions) - because even when mounting my SQLite database into `/home/app/db/ShopAndEat.db` as a volume, I had to `chown` the database on the host with the app's UID because otherwise, writing into the DB failed with `Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 8: 'attempt to write a readonly database'.`

**Reply by mu88** [reply]
Dear @baronfel - as I really appreciate your valuable feedback during your precious time, would you prefer to split my questions, e. g. like this?
1. File a dedicated issue in the Playwright repo regarding rootless containers.
2. Discuss the `Dockerfile` and SQLite issues here.

Whatever fits you best :)

**Reply by baronfel** [reply]
yeah, that seems like a good split to me - based on what you've shown there will need to be a larger discussion with Playwright about how to use it in such an environment.

regarding the database permissions issue, that's interesting - I'd be very interested in what the permissions to the mounted database location were when seen from inside the container - just because `/home/app` had the correct permissions it could be the case that the Docker-side volume mount had incorrect permissions?

**Reply by mu88** [reply]
> regarding the database permissions issue, that's interesting - I'd be very interested in what the permissions to the mounted database location were when seen from inside the container - just because /home/app had the correct permissions it could be the case that the Docker-side volume mount had incorrect permissions?

I'm not a Docker expert, but here's what I tried:
1. `mkdir ~/Docker/tiyl-test`
2. `docker run -d -v /home/myUser/Docker/tiyl-test/:/home/app/data -p:5555:8080 --name tiyl-test mu88/thisisyourlife:latest`
3. `docker logs tiyl-test`

Which gives me:
```
[16:02:47 INF] Creating database directory `/home/app/data/db`
[16:02:47 ERR] An error occurred creating the DB
System.UnauthorizedAccessException: Access to the path '/home/app/data/db' is denied.
 ---> System.IO.IOException: Permission denied
   --- End of inner exception stack trace ---
   at System.IO.FileSystem.CreateDirectory(String fullPath, UnixFileMode unixCreateMode)
   at System.IO.Directory.CreateDirectory(String path)
   at Persistence.FileSystem.CreateDirectory(String path) in /home/runner/work/ThisIsYourLife/ThisIsYourLife/src/Persistence/Impl/FileSystem.cs:line 26
   at Persistence.Storage.EnsureStorageExistsAsync() in /home/runner/work/ThisIsYourLife/ThisIsYourLife/src/Persistence/Impl/Storage.cs:line 82
   at Program.<<Main>$>g__CreateDbIfNotExistsAsync|0_4(IHost host) in /home/runner/work/ThisIsYourLife/ThisIsYourLife/src/WebApp/Program.cs:line 68
```

Running `ls -lh` gives me:
```
drwxr-xr-x  2 myUser myUser 4.0K Nov 28 17:02 tiyl-test
```

How can I check the permissions you're talking about?

**Reply by baronfel** [reply]
To check the permissions of the directory _inside the running container_ you'd need to launch the container with a different entrypoint and check the permissions. That would look something like this:

```terminal
$ docker run -it --rm -v /home/myUser/Docker/tiyl-test/:/home/app/data -p:5555:8080 --entrypoint /bin/sh --name tiyl-test mu88/thisisyourlife:latest
> ls -la /home/app/data
```

**Reply by mu88** [reply]
That gives the following:
```
$ ls -la /home/app/data
total 8
drwxr-xr-x 2 1001 1001 4096 Nov 28 16:02 .
drwxr-xr-x 1 app  app  4096 Nov 29 06:54 ..
```

**Reply by baronfel** [reply]
Ok, that's a clue! It looks like the ownership of the `data` directory is 1001:1001 instead of app:app.

The bad news is this - there doesn't seem to be a way to set the permissions on a volume that's mounted from the Docker CLI: https://stackoverflow.com/a/56990338

Answers suggest creating a volume inside the container during container creation, but unfortunately we don't support Volumes in the SDK tech.

So I think you may be in a bit of a wall with the rootless execution at the moment. We can look at implementing Volume support as it's clear to me that is a key part of the overall rootless end to end story.

**Reply by mu88** [reply]
Thx for your explanations - I'm learning a lot :)

I agree that volumes are an essential part, therefore having support for that in the SDK would be great. Do you have an issue for me that I can subscribe to? With that, I think we could close this discussion.

**Reply by baronfel** [reply, marked as answer]
We added one just over a week ago here: https://github.com/dotnet/sdk-container-builds/issues/526

I'm going to add a comment to this issue based on our discussion here right now!
